byzcoin/contract_deferred.go

```diff
@@ -0,0 +1,404 @@
+
+		// In the following we are creating a new StagingStateTrie from the
+		// readonly state try by copying the data.
+		nonce, err2 := rst.GetNonce()
+		if err2 != nil {
+			return nil, nil, errors.New("couldn't get the nonce: " + err2.Error())
+		}
+		sst, err2 := newMemStagingStateTrie(nonce)
+		if err2 != nil {
+			return nil, nil, errors.New("Failed to created stagingStateTrie: " + err2.Error())
+		}
+		rst.ForEach(sst.Set)
+
+		instructionIDs := make([][]byte, len(c.DeferredData.ProposedTransaction.Instructions))
+
+		for i, proposedInstr := range c.DeferredData.ProposedTransaction.Instructions {
+
+			// In case it goes well, we want to return the proposed Tx InstanceID
+			instructionIDs[i] = proposedInstr.DeriveID("").Slice()
+
+			instructionType := proposedInstr.GetType()
+
+			var contractID string
+			switch instructionType {
+			case SpawnType:
+				contractID = proposedInstr.Spawn.ContractID
+			case InvokeType:
+				contractID = proposedInstr.Invoke.ContractID
+			case DeleteType:
+				contractID = proposedInstr.Delete.ContractID
+			}
+
+			fn, exists := c.s.GetContractConstructor(contractID)
+			if !exists {
+				return nil, nil, errors.New("couldn't get the root function")
+			}
+			rootInstructionBuff, err := protobuf.Encode(&proposedInstr)
+			if err != nil {
+				return nil, nil, errors.New("couldn't encode the root instruction buffer")
+			}
+			contract, err := fn(rootInstructionBuff)
+			if err != nil {
+				return nil, nil, errors.New("couldn't get the root contract")
+			}
+			err = contract.VerifyDeferredInstruction(sst, proposedInstr, c.DeferredData.InstructionHashes[i])
+			if err != nil {
+				return nil, nil, fmt.Errorf("verifying the instruction failed: %s", err)
+			}
+
+			var stateChanges []StateChange
+			switch instructionType {
+			case SpawnType:
+				stateChanges, _, err = contract.Spawn(sst, proposedInstr, coins)
+			case InvokeType:
+				stateChanges, _, err = contract.Invoke(sst, proposedInstr, coins)
+			case DeleteType:
+				stateChanges, _, err = contract.Delete(sst, proposedInstr, coins)
+
+			}
+
+			sst.StoreAll(stateChanges)
+
+			if err != nil {
+				return nil, nil, fmt.Errorf("error while executing an instruction: %s", err)
+			}
+			sc = append(sc, stateChanges...)
+
+		}
+
+		c.DeferredData.ExecResult = instructionIDs
+		// At this stage all verification passed. We can then decrease the
+		// NumExecution counter.
+		c.DeferredData.NumExecution = c.DeferredData.NumExecution - 1
+		resultBuf, err2 := protobuf.Encode(&c.DeferredData)
+		if err2 != nil {
+			return nil, nil, errors.New("couldn't encode the result")
+		}
+		sc = append(sc, NewStateChange(Update, inst.InstanceID,
+			ContractDeferredID, resultBuf, darcID))
+
+		return
+	default:
+		return nil, nil, errors.New("deferred contract can only addProof and execProposedTx")
+	}
+}
+
+func (c *contractDeferred) Delete(rst ReadOnlyStateTrie, inst Instruction, coins []Coin) (sc []StateChange, cout []Coin, err error) {
```

Comment:
let me know if i missed it, but i feel there's a missing feature surrounding the delete
during our meeting with @jeffallen we decided to add a feature that allows anybody to delete the instance if has expired, but i didn't find it in this PR. actually, it might be better if it's in another PR. what might be possible is the following: add an optional field in `BasicContract` called `ExpirationBlockIndex`. by default it's 0 and it's not checked. but if it is set, then `VerifyInstruction` will check whether it is expired and if it is, anybody can delete it.

Reply:
This is implemented inside the `VerifyInstruction` function, where I allow anyone to delete the contract if it has expired.